### PR TITLE
Fix security issues in copilot chat app

### PR DESCRIPTION
**Pull request by DavidParks8**
### Description
api key auth was removed because it was not possible to be secure in a user context. Any user was able to operate on any other user's chats. A new auth policy was introduced to validate userIds coming from tokens rather than request bodies.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: samples/apps/copilot-chat-app/webapi/Controllers/ProbeController.cs
Comment (DavidParks8) [inline samples/apps/copilot-chat-app/webapi/Controllers/ProbeController.cs:1]: This file was replaced by asp.net core health checks.

File: samples/apps/copilot-chat-app/webapi/Auth/PassThroughAuthenticationHandler.cs [hunk 1]
```diff
@@ -15,6 +16,8 @@ namespace SemanticKernel.Service.Auth;
 public class PassThroughAuthenticationHandler : AuthenticationHandler<AuthenticationSchemeOptions>
 {
     public const string AuthenticationScheme = "PassThrough";
+    private const string DefaultUserId = "c05c61eb-65e4-4223-915a-fe72b0c9ece1";
```
Comment (DavidParks8) [inline samples/apps/copilot-chat-app/webapi/Auth/PassThroughAuthenticationHandler.cs:19]: There is nothing special about this guid. It just needs to be the same for each request.

File: samples/apps/copilot-chat-app/webapi/Controllers/SemanticKernelController.cs [hunk 2]
```diff
@@ -34,34 +36,45 @@ public SemanticKernelController(ILogger<SemanticKernelController> logger)
...
     /// </remarks>
     /// <param name="kernel">Semantic kernel obtained through dependency injection</param>
+    /// <param name="askConverter">Converter to use for converting Asks.</param>
+    /// <param name="chatSessionRepository">Storage for chat sessions.</param>
+    /// <param name="authInfo">Authenticated info about the user for the current request.</param>
     /// <param name="ask">Prompt along with its parameters</param>
     /// <param name="skillName">Skill in which function to invoke resides</param>
     /// <param name="functionName">Name of function to invoke</param>
     /// <returns>Results consisting of text generated by invoked function along with the variable in the SK that generated it</returns>
-    [Authorize]
```
Comment (DavidParks8) [inline samples/apps/copilot-chat-app/webapi/Controllers/SemanticKernelController.cs:41]: All endpoints now require authorization by default..

File: samples/apps/copilot-chat-app/webapp/src/components/views/BackendProbe.tsx [hunk 3]
```diff
@@ -10,7 +10,7 @@ interface IData {
 
 const BackendProbe: FC<IData> = ({ uri, onBackendFound }) => {
     useEffect(() => {
-        const requestUrl = new URL('probe', uri);
+        const requestUrl = new URL('healthz', uri);
```
Comment (DavidParks8) [inline samples/apps/copilot-chat-app/webapp/src/components/views/BackendProbe.tsx:13]: This now follows the recommended endpoint from the healthchecks docs.

File: samples/apps/copilot-chat-app/webapp/src/libs/services/ChatService.ts [hunk 4]
```diff
@@ -9,22 +9,18 @@ import { BaseService } from './BaseService';
...
-            userId: userId,
-            userName: userName,
-            title: title,
+            title,
         };
 
         const result = await this.getResponseAsync<IChatSession>(
             {
-                commandPath: 'chatSession/create',
+                commandPath: 'chatSessions',
```
Comment (DavidParks8) [inline samples/apps/copilot-chat-app/webapp/src/libs/services/ChatService.ts:21]: Chat sessions urls were altered to more closely follow microsoft's api design standards.

File: samples/apps/copilot-chat-app/webapp/src/libs/useChat.ts [hunk 5]
```diff
@@ -101,14 +99,6 @@ export const useChat = () => {
         const ask = {
             input: value,
             variables: [
-                {
-                    key: 'userId',
```
Comment (DavidParks8) [inline samples/apps/copilot-chat-app/webapp/src/libs/useChat.ts:105]: These variables are now injected server-side.

File: samples/apps/copilot-chat-app/webapi/ServiceExtensions.cs [hunk 6]
```diff
@@ -78,39 +85,50 @@ internal static IServiceCollection AddCors(this IServiceCollection services)
     }
 
     /// <summary>
-    /// Add authorization services
+    /// Add authentication services
     /// </summary>
-    internal static IServiceCollection AddAuthorization(this IServiceCollection services, IConfiguration configuration)
+    internal static IServiceCollection AddCopilotChatAuthentication(this IServiceCollection services, IConfiguration configuration)
```
Comment (DavidParks8) [inline samples/apps/copilot-chat-app/webapi/ServiceExtensions.cs:90]: These methods were renamed to include "CopilotChat" in order to make it more obvious that they are customizations and not the .net Add* methods.

File: samples/apps/copilot-chat-app/webapi/CopilotChat/Controllers/DocumentImportController.cs [hunk 7]
```diff
@@ -86,12 +85,18 @@ private enum SupportedFileType
...
-        if (documentImportForm.DocumentScope == DocumentImportForm.DocumentScopes.Chat
-            && !(await this.UserHasAccessToChatAsync(documentImportForm.UserId, documentImportForm.ChatId)))
+        var chatSession = await this._chatSessionRepository.FindByIdAsync(documentImportForm.ChatId.ToString());
+        if (chatSession == null)
         {
-            return this.BadRequest("User does not have access to the chat session.");
+            return this.NotFound("Session does not exist.");
         }
 
+        if (documentImportForm.DocumentScope == DocumentImportForm.DocumentScopes.Chat && authInfo.UserId != chatSession.UserId)
```
Comment (DavidParks8) [inline samples/apps/copilot-chat-app/webapi/CopilotChat/Controllers/DocumentImportController.cs:94]: This is still a problem because any user can upload a document for all users without any validation, but perhaps that was by design?

File: samples/apps/copilot-chat-app/webapi/CopilotChat/Controllers/ChatHistoryController.cs [hunk 8]
```diff
@@ -54,14 +58,14 @@ public class ChatHistoryController : ControllerBase
     /// <param name="chatParameters">Object that contains the parameters to create a new chat.</param>
     /// <returns>The HTTP action result.</returns>
     [HttpPost]
```
Comment (DavidParks8) [inline samples/apps/copilot-chat-app/webapi/CopilotChat/Controllers/ChatHistoryController.cs:60]: This should be HttpPut, but that is outside the scope of the PR.

File: samples/apps/copilot-chat-app/webapi/CopilotChat/Controllers/DocumentImportController.cs (hunk 7 above)
Comment (TaoChenOSU) [reply samples/apps/copilot-chat-app/webapi/CopilotChat/Controllers/DocumentImportController.cs:94]: Yes, this is by design

File: samples/apps/copilot-chat-app/webapi/CopilotChat/Controllers/DocumentImportController.cs [hunk 9]
```diff
@@ -117,7 +122,7 @@ private enum SupportedFileType
             return this.BadRequest(ex.Message);
         }
 
-        return this.Ok();
+        return this.Accepted();
```
Comment (TaoChenOSU) [inline samples/apps/copilot-chat-app/webapi/CopilotChat/Controllers/DocumentImportController.cs:125]: Should probably not use Accepted() since Accepted means the operation may not have been completed yet.

File: samples/apps/copilot-chat-app/webapi/CopilotChat/Controllers/DocumentImportController.cs [hunk 10]
```diff
@@ -86,12 +85,18 @@ private enum SupportedFileType
             return this.BadRequest("File size exceeds the limit.");
         }
 
-        if (documentImportForm.DocumentScope == DocumentImportForm.DocumentScopes.Chat
-            && !(await this.UserHasAccessToChatAsync(documentImportForm.UserId, documentImportForm.ChatId)))
+        var chatSession = await this._chatSessionRepository.FindByIdAsync(documentImportForm.ChatId.ToString());
```
Comment (TaoChenOSU) [inline samples/apps/copilot-chat-app/webapi/CopilotChat/Controllers/DocumentImportController.cs:88]: The ChatId in the documentImportForm will only be valid when the scope is Chat. When the scope is Global, the document will be available to all chat. Please do not perform the validation here.

File: samples/apps/copilot-chat-app/webapi/ServiceExtensions.cs [hunk 11]
```diff
@@ -38,8 +40,8 @@ internal static IServiceCollection AddOptions(this IServiceCollection services,
         var foo = services.BuildServiceProvider().GetService<IOptions<AIServiceOptions>>();
 
         // Authorization configuration
-        services.AddOptions<AuthorizationOptions>()
-            .Bind(configuration.GetSection(AuthorizationOptions.PropertyName))
+        services.AddOptions<ChatAuthenticationOptions>()
```
Comment (TaoChenOSU) [inline samples/apps/copilot-chat-app/webapi/ServiceExtensions.cs:43]: All Chat related logic should only live under CopilotChat/

Comment (TaoChenOSU) [reply samples/apps/copilot-chat-app/webapi/ServiceExtensions.cs:43]: Under CopilotChat/, there is an Extenstions module.

**Comment by TaoChenOSU** [comment]
Thanks for your PR!!

Just a note for our team, this PR will create a lot of conflicts with the multi-user feature branch. In the multi-user setting, some chat-related authentication logic may need to change from this PR. For example, the data schema for ChatSession will be different in the multi-user setting. We need to talk about which one we want to let in first.

**Comment by DavidParks8** [comment]
This PR needs to go in first because if it had been filed via MSRC, your team would have been required to drop everything to fix it immediately. Security takes precedence over new features. I chose to instead save your team all that trouble and, with the goahead from adrian, implement it.

File: samples/apps/copilot-chat-app/webapi/Controllers/SemanticKernelController.cs [hunk 12]
```diff
@@ -34,34 +36,45 @@ public SemanticKernelController(ILogger<SemanticKernelController> logger)
...
     [Route("skills/{skillName}/functions/{functionName}/invoke")]
     [HttpPost]
     [ProducesResponseType(StatusCodes.Status200OK)]
     [ProducesResponseType(StatusCodes.Status400BadRequest)]
     [ProducesResponseType(StatusCodes.Status404NotFound)]
     public async Task<ActionResult<AskResult>> InvokeFunctionAsync(
```
[the remark this answers is not in the document]
Comment (DavidParks8) [reply samples/apps/copilot-chat-app/webapi/Controllers/SemanticKernelController.cs:51]: Yet, it must contain the minimum components required to validate that the user is allowed to pass specific context variables. Otherwise, users will be able to invoke skills for chats they do not own.

File: samples/apps/copilot-chat-app/webapi/CopilotChat/Controllers/ChatController.cs [hunk 13]
```diff
@@ -65,18 +69,31 @@ public ChatController(ILogger<ChatController> logger)
...
         [FromHeader] OpenApiSkillsAuthHeaders openApiSkillsAuthHeaders)
     {
         this._logger.LogDebug("Chat request received.");
-
-        // Put ask's variables in the context we will use.
-        var contextVariables = new ContextVariables(ask.Input);
-        foreach (var input in ask.Variables)
+        const string chatIdKey = "chatId";
+        var chatIdFromContext = ask.Variables.FirstOrDefault(x => x.Key == chatIdKey);
+        if (chatIdFromContext.Key is chatIdKey)
```
[the remark this answers is not in the document]
Comment (DavidParks8) [reply samples/apps/copilot-chat-app/webapi/CopilotChat/Controllers/ChatController.cs:81]: Doesn't seem like an issue. If it doesn't contain the chat id key, then the if block will be skipped.

File: samples/apps/copilot-chat-app/webapi/Models/Ask.cs [hunk 14]
```diff
@@ -1,12 +1,15 @@
...
 using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
 using System.Linq;
+using SemanticKernel.Service.Options;
 
 namespace SemanticKernel.Service.Models;
 
 public class Ask
 {
+    [Required, NotEmptyOrWhitespace]
```
[the remark this answers is not in the document]
Comment (DavidParks8) [reply samples/apps/copilot-chat-app/webapi/Models/Ask.cs:12]: It needs to be required in the chat controller, but not in the sk controller. I split it out to a separate model for validation.

File: samples/apps/copilot-chat-app/webapi/CopilotChat/Controllers/DocumentImportController.cs (hunk 10 above)
Comment (DavidParks8) [reply samples/apps/copilot-chat-app/webapi/CopilotChat/Controllers/DocumentImportController.cs]: Fixed to only do validation when the mode is chat.

**Comment by adrianwyatt** [comment]
> @TaoChenOSU, This PR needs to go in first because if it had been filed via MSRC, your team would have been required to drop everything to fix it immediately. Security takes precedence over new features. I chose to instead save your team all that trouble and, with the goahead from @adrianwyatt, implement it.

I am looking at this today and the review will likely bleed into tomorrow. Thank you for your awesome work on this!

File: samples/apps/copilot-chat-app/webapi/CopilotChat/Controllers/ChatController.cs [hunk 15]
```diff
@@ -62,18 +66,31 @@ public ChatController(ILogger<ChatController> logger)
     public async Task<IActionResult> ChatAsync(
         [FromServices] IKernel kernel,
         [FromServices] CopilotChatPlanner planner,
-        [FromBody] Ask ask,
+        [FromServices] AskConverter askConverter,
```
Comment (TaoChenOSU) [inline samples/apps/copilot-chat-app/webapi/CopilotChat/Controllers/ChatController.cs:69]: I think we don't necessarily need to use the `Ask` for this API since it only invokes the `ChatSkill`. We can create a new model for the body that contains the ChatId and the message. Would love @adrianwyatt's take on this.

File: samples/apps/copilot-chat-app/webapi/Controllers/SemanticKernelController.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (TaoChenOSU) [reply samples/apps/copilot-chat-app/webapi/Controllers/SemanticKernelController.cs:51]: This controller will not be used for any chat related features. Maybe we need to be explicit about it.

File: samples/apps/copilot-chat-app/webapi/Utilities/AskConverter.cs [hunk 16]
```diff
@@ -0,0 +1,38 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
```
[the remark this answers is not in the document]
Comment (DavidParks8) [reply samples/apps/copilot-chat-app/webapi/Utilities/AskConverter.cs:1]: It isn't. It is used in the sk controller as well.

File: samples/apps/copilot-chat-app/webapi/CopilotChat/Controllers/ChatController.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (DavidParks8) [reply samples/apps/copilot-chat-app/webapi/CopilotChat/Controllers/ChatController.cs:78]: This would be net new validation. That is fine, but is more suited to its own PR.

File: samples/apps/copilot-chat-app/webapi/Options/ChatAuthenticationOptions.cs [hunk 17]
```diff
@@ -7,33 +7,26 @@ namespace SemanticKernel.Service.Options;
 /// <summary>
```
[the remark this answers is not in the document]
Comment (DavidParks8) [reply samples/apps/copilot-chat-app/webapi/Options/ChatAuthenticationOptions.cs:7]: Such segregation is not recommended or encouraged due to the associated risks.

By separating the Copilot chat components from the generic sk service code, it opens up the possibility of invoking any registered skill without proper validation. This could potentially lead to unauthorized access and misuse of the system. Maintaining a cohesive codebase ensures that necessary checks and balances are in place to validate and authorize skill invocations, thus safeguarding the integrity of the system.

Instead of segregating the components, I would suggest focusing on implementing robust security and validation mechanisms within the codebase. This way, you can ensure that only authorized and validated skills can be invoked, mitigating the risk of potential vulnerabilities. It is crucial to prioritize system security and maintain a unified codebase for better control and protection.

If you were to bring this to a security review, I would tell you to delete the sk controller entirely.

File: samples/apps/copilot-chat-app/webapi/Controllers/SemanticKernelController.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (DavidParks8) [reply samples/apps/copilot-chat-app/webapi/Controllers/SemanticKernelController.cs:51]: Regardless of what it will be used for in the app, it still allows for remote execution of the chat skill if a bad actor constructs the proper request. This must be guarded against.

File: samples/apps/copilot-chat-app/webapi/CopilotChat/Controllers/ChatController.cs (hunk 15 above)
Comment (DavidParks8) [reply samples/apps/copilot-chat-app/webapi/CopilotChat/Controllers/ChatController.cs:69]: I agree, but such a change is outside the intended scope of this PR. The goal of the PR is only to close the most critical security gap.

File: samples/apps/copilot-chat-app/webapi/Controllers/SemanticKernelController.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (TaoChenOSU) [reply samples/apps/copilot-chat-app/webapi/Controllers/SemanticKernelController.cs:51]: Totally agreed! Need further discussion with @adrianwyatt and @hathind-ms

**Comment by TaoChenOSU** [comment]
Is there anything additional I need to perform before launching the web API and the web app besides the current instructions documented in the ReadMe? The web API throws errors when the web app tries to retrieve chat sessions. I have tried both pass through auth and AAD auth.

**Comment by adrianwyatt** [comment]
Thanks again for the PR and especially outlining the concerns. The team is meeting this week to discuss how to mitigate the issues in line with our near-future planning. We'll be referencing this PR and will send you the plan as an FYI as soon as we have one.

**Comment by adrianwyatt** [comment]
@DavidParks8 and I had a great real-time conversation and we'll be bringing in these changes piecewise over the next week or two. Closing this one in light of a set of incoming PRs.
